interfaces: login-session-control: add further D-Bus interfaces

IsaacJT opened the pull request:
These new D-Bus interfaces allow the snaps to additionally control login sessions. This includes the ability to terminate or kill login sessions. These new interfaces were obtained from https://www.freedesktop.org/software/systemd/man/org.freedesktop.login1.html.

The unit tests have also been extended to add a check to ensure that the D-Bus rules are correctly included in the AppArmor snippet.

IsaacJT commented:
Functional test:

Prior to commit: 

```
# dbus-send --system --print-reply --dest=org.freedesktop.login1 /org/freedesktop/login1 'org.freedesktop.login1.Manager.ListSessions'
Error org.freedesktop.DBus.Error.AccessDenied: An AppArmor policy prevents this sender from sending this message to this recipient; type="method_call", sender=":1.20" (uid=0 pid=1784 comm="dbus-send --system --print-reply --dest=org.freede" label="snap.login-test.login-test (enforce)") interface="org.freedesktop.login1.Manager" member="ListSessions" error name="(unset)" requested_reply="0" destination="org.freedesktop.login1" (uid=0 pid=783 comm="/lib/systemd/systemd-logind " label="unconfined")
```

Using snapd with the new commit:

```
# dbus-send --system --print-reply --dest=org.freedesktop.login1 /org/freedesktop/login1 'org.freedesktop.login1.Manager.ListSessions'
method return time=1662379401.351886 sender=:1.5 -> destination=:1.24 serial=106 reply_serial=2
   array [
      struct {
         string "5"
         uint32 1000
         string "iotuc"
         string ""
         object path "/org/freedesktop/login1/session/_35"
      }
      struct {
         string "3"
         uint32 1000
         string "iotuc"
         string ""
         object path "/org/freedesktop/login1/session/_33"
      }
   ]
```

Terminating sessions using the `org.freedesktop.login1.Manager.TerminateSession` also works.

alexmurray reviewed (changes requested):
Thanks - so a couple things - can some rationale please be added to the PR so we can understand the use-case for these new APIs? Also the title of the PR needs to be changed so that the rest of the tests get run as currently they are blocked on the PR name check test.

IsaacJT commented:
Hi @alexmurray,

I've already adjusted the name so those tests should pass now.

We have a customer who would like to be able to terminate active SSH sessions on Ubuntu Core (why exactly I'm not sure, but that's their wish. I think it has something to do with ensuring that no one else is connected to the device during specific application updates) and to do that they need access to the ListSessions and TerminateSession D-Bus interfaces. I added the rest of the *Session interfaces for the sake of completeness.

alexmurray reviewed (commented):
Thanks for the additional info - I wonder should the `org.freedesktop.login1.Session` interface permissions also be updated to add the equivalent `Kill` and `Terminate` methods as well?

mardy commented:
> `org.freedesktop.login1.Session`

I actually wonder, why do we have a list of allowed methods, instead of just granting access to the whole Session interface? After all, this snapd interface is called `login-session-control` and is not autoconnecting, so I think it's expected that one who has access to this interface should be able to control the session in full.

I think it's a good idea to keep the member list for the Manager object, since that also includes methods like `Poweroff` and `Reboot`, but at least for `Session` and `Seat` I don't see a problem.

IsaacJT commented:
@mardy something like this? (I'm not sure if * works in the AA profile...)

```
diff --git a/interfaces/builtin/login_session_control.go b/interfaces/builtin/login_session_control.go
index b6fea480f0..df643c44f4 100644
--- a/interfaces/builtin/login_session_control.go
+++ b/interfaces/builtin/login_session_control.go
@@ -38,21 +38,21 @@ dbus (send,receive)
     bus=system
     path=/org/freedesktop/login1/{seat,session}/**
     interface=org.freedesktop.DBus.Properties
-    member={GetAll,PropertiesChanged,Get}
+    member=*
     peer=(label=unconfined),

 dbus (send,receive)
     bus=system
     path=/org/freedesktop/login1/seat/**
     interface=org.freedesktop.login1.Seat
-    member={ActiveSession,SwitchTo}
+    member=*
     peer=(label=unconfined),

 dbus (send,receive)
     bus=system
     path=/org/freedesktop/login1/session/**
     interface=org.freedesktop.login1.Session
-    member={TakeControl,TakeDevice,PauseDevice,PauseDeviceComplete,ResumeDevice,ReleaseDevice,Active,State,Lock,Unlock,Activate,ReleaseControl}
+    member=*
     peer=(label=unconfined),

 dbus (send,receive)
```

alexmurray commented:
> @mardy something like this? (I'm not sure if * works in the AA profile...)

No - just drop the `member=` part of the rules and that will provide access to all members by default.

mardy reviewed (approved):
This LGTM, thanks for the updates! But please try to get an additional approval of another developer before merging this.

alexmurray commented on interfaces/builtin/login_session_control.go:
```diff
@@ -38,28 +38,25 @@ dbus (send,receive)
     bus=system
     path=/org/freedesktop/login1/{seat,session}/**
     interface=org.freedesktop.DBus.Properties
-    member={GetAll,PropertiesChanged,Get}
```
Please do not remove this line as it then allows to `Set` the various properties directly.

mardy replied:
OTOH, this is only for the `seat` and `session` objects (judging from their path), so I guess the client should be able to alter them?

alexmurray replied:
Perhaps - I am just worried that modifying the properties directly may not be safe, and that it is better to use the various methods of the interfaces to do this instead.

IsaacJT replied:
Maybe we can leave it out of this PR and include it if there's a use case in the future?

mardy replied:
Sounds good to me!

IsaacJT replied:
Done!

alexmurray reviewed (changes requested):
Please restore the `member` line for the `interface=org.freedesktop.DBus.Properties` - otherwise LGTM.

alexmurray reviewed (approved):
LGTM!

mvo5 reviewed (approved):
Thank you
